File: ployglots/mta-ployglot-pkg-app-yaml-loader/packages/py/src/app_yaml_loader/loader.py
"""YAML loader — read + parse + return ordered map keyed by absolute path."""

from __future__ import annotations

import logging
import os
import os.path
from collections import OrderedDict
from typing import Any, Iterable, Literal

import yaml

from . import _io
from .cache import cache_get, cache_set, is_miss
from .errors import LoadError
from .paths import build_config_files, resolve_app_env, resolve_config_dir

_VALID_MISSING = ("raise", "skip")

_DEFAULT_LOGGER = logging.getLogger("app_yaml_loader")


def load_files(
    paths: Iterable[str],
    *,
    missing: Literal["raise", "skip"] = "raise",
    force: bool = False,
    logger: logging.Logger | None = None,
) -> "OrderedDict[str, Any]":
    if missing not in _VALID_MISSING:
        raise ValueError(f"Invalid missing strategy: {missing}")

    log = logger if logger is not None else _DEFAULT_LOGGER

    out: "OrderedDict[str, Any]" = OrderedDict()

    for p in paths:
        if not isinstance(p, str):
            raise TypeError(f"paths[*] must be str, got {type(p).__name__}")
        abs_path = os.path.abspath(p)

        if not force:
            hit = cache_get(abs_path)
            if not is_miss(hit):
                out[abs_path] = hit
                continue

        try:
            raw = _io.read_text(abs_path)
        except FileNotFoundError as e:
            if missing == "skip":
                log.warning("app-yaml-loader: skipping missing file %s", abs_path)
                continue
            raise LoadError(f"Failed to read {abs_path}", path=abs_path, cause=e) from e
        except OSError as e:
            raise LoadError(f"Failed to read {abs_path}", path=abs_path, cause=e) from e

        try:
            parsed = yaml.safe_load(raw)
        except yaml.YAMLError as e:
            raise LoadError(f"Failed to parse {abs_path}", path=abs_path, cause=e) from e

        if parsed is None:
            parsed = {}

        cache_set(abs_path, parsed)
        out[abs_path] = cache_get(abs_path)

    return out


def load_from_config_dir(
    *,
    config_dir: str | None = None,
    caller_dir: str | None = None,
    app_env: str | None = None,
    base_files: list[str] | tuple[str, ...] | None = None,
    env_suffixes: list[str] | tuple[str, ...] | None = None,
    missing: Literal["raise", "skip"] = "raise",
    force: bool = False,
    logger: logging.Logger | None = None,
) -> "OrderedDict[str, Any]":
    resolved_dir = resolve_config_dir(config_dir, caller_dir)
    resolved_env = resolve_app_env(app_env)
    files = build_config_files(resolved_dir, resolved_env, base_files, env_suffixes)
    return load_files(files, missing=missing, force=force, logger=logger)

File: ployglots/mta-ployglot-pkg-app-yaml-loader/packages/mjs/src/loader.mjs
import * as path from 'node:path';
import yaml from 'js-yaml';
import { io } from './_io.mjs';
import { LoadError } from './errors.mjs';
import { resolveConfigDir, resolveAppEnv, buildConfigFiles } from './paths.mjs';
import { cacheGet, cacheSet } from './cache.mjs';

const VALID_MISSING = new Set(['raise', 'skip']);

export async function loadFiles(paths, { missing = 'raise', force = false, logger } = {}) {
  if (!VALID_MISSING.has(missing)) {
    throw new Error(`Invalid missing strategy: ${missing}`);
  }

  const out = new Map();

  for (const p of paths) {
    if (typeof p !== 'string') {
      throw new TypeError(`paths[*] must be string, got ${typeof p}`);
    }
    const abs = path.resolve(p);

    if (!force) {
      const hit = cacheGet(abs);
      if (hit !== undefined) {
        out.set(abs, hit);
        continue;
      }
    }

    let raw;
    try {
      raw = await io.readFile(abs, 'utf8');
    } catch (e) {
      if (e && e.code === 'ENOENT' && missing === 'skip') {
        if (logger && typeof logger.warn === 'function') {
          logger.warn(`app-yaml-loader: skipping missing file ${abs}`);
        }
        continue;
      }
      throw new LoadError(`Failed to read ${abs}`, { path: abs, cause: e });
    }

    let parsed;
    try {
      parsed = yaml.load(raw) ?? {};
    } catch (e) {
      throw new LoadError(`Failed to parse ${abs}`, { path: abs, cause: e });
    }

    cacheSet(abs, parsed);
    // cacheGet returns a fresh clone; use it so the caller's map mirrors cache semantics.
    out.set(abs, cacheGet(abs));
  }

  return out;
}

export async function loadFromConfigDir(opts = {}) {
  const configDir = resolveConfigDir(opts.configDir, opts.callerDir);
  const appEnv = resolveAppEnv(opts.appEnv);
  const files = buildConfigFiles(configDir, appEnv, opts.baseFiles, opts.envSuffixes);
  return loadFiles(files, { missing: opts.missing, force: opts.force, logger: opts.logger });
}
